### PR TITLE
[MM-53256] Use buildinfo instead of env variables for mmctl

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
Since go1.18 [`debug.BuildSetting`](https://pkg.go.dev/runtime/debug@go1.18beta2#BuildSetting) is provided as part of [`debug.BuildInfo`](https://pkg.go.dev/runtime/debug@go1.18beta2#BuildInfo). It contains build information like the build commit or build time. 

This PR makes use of the build information and ditches some custom `Makefile` code. The main advantage is that every build of `mmclt` dow contains the correct version information.

#### Old Output (`go install github.com/mattermost/mmctl/v6@master`)

```
mmctl:
Version:	unspecified
GitCommit:	dev mode
GitTreeState:	unknown
BuildDate:	unknown
GoVersion:	go1.20.4
Compiler:	gc
Platform:	linux/amd64
```

#### New Output (`make mmctl-build`)
```
mmctl:
Version:	8.0.0
BuiltDate:	2023-06-28T10:52:41Z
CommitDate:	2023-06-28T10:52:31Z
GitCommit:	cd7d3f2a44cb5177180f23d14aac950d261831c1
GitTreeState:	clean
GoVersion:	go1.20.4
Compiler:	gc
Platform:	amd64/linux
```

#### New Output (`go install ./cmd/mmctl`)
```
mmctl:
Version:	8.0.0
BuiltDate:	dev
CommitDate:	2023-06-28T10:52:31Z
GitCommit:	cd7d3f2a44cb5177180f23d14aac950d261831c1
GitTreeState:	clean
GoVersion:	go1.20.4
Compiler:	gc
Platform:	amd64/linux

```

I also tried the same approach with the server binary, but it didn't provide much value, as the EE hash also needs to be included. I'm 0/5 on whether it's useful. You can find the diff via https://github.com/mattermost/mattermost/compare/buildinfo_mmctl...buildinfo_server.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53256

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/cmd/mmctl/commands/version.go [hunk 1]
```diff
@@ -51,14 +50,50 @@ type Info struct {
...
+
+		os       string
+		arch     string
+		compiler string
+	)
+
+	for _, s := range info.Settings {
+		switch s.Key {
+		case "vcs.revision":
+			revision = s.Value
```
Comment (isacikgoz) [inline server/cmd/mmctl/commands/version.go:72]: Does this get the latest tag? If you install `master`, would it register version as the latest tag? (in your example it looks like `v7.11` hence I'm asking)

Comment (hanzei) [reply server/cmd/mmctl/commands/version.go:72]: `revision` is always the HEAD commit sha. Maybe you are referring to `Version`, which is the latest version listed in `model/versio.go`? That is curtently v7.11.0 (and need to get updated seperatetly).

Comment (isacikgoz) [reply server/cmd/mmctl/commands/version.go:72]: Right, thanks for the clarification.

**Review by isacikgoz — APPROVED** [review]

File: server/cmd/mmctl/commands/version.go [hunk 2]
```diff
@@ -51,14 +50,50 @@ type Info struct {
...
+		case "vcs.time":
+			buildTime = s.Value
```
[the remark this answers is not in the document]
Comment (agnivade) [reply server/cmd/mmctl/commands/version.go:74]: IMO, we would be digressing from expectations. To me, "build date" means the date when the build was made. Not the commit date. We also use that to see when was Community last built. Although this is `mmctl` and not server, but I'd like to keep the semantics same here.

If you want to preserve the "commit time", I'd be fine with another field for that.

**Review by agnivade — CHANGES_REQUESTED** [review]
One minor thing to note is that this info is _only_ available while running the binary, and not via "go run". This was a bit unexpected to me as I'd expect the output of a CLI via "go run" to be the same as executing the binary.

It's fine to make the change for mmctl, but I'd be careful to change this for MM.

**Review by mgdelacroix — APPROVED** [review]
Nice improvement! LGTM besides @agnivade's comment on build time :+1:

File: server/cmd/mmctl/commands/version.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/cmd/mmctl/commands/version.go]: I've renamed "build date" to "commit date" and reintroduced "build date" as it was before this PR.

File: server/Makefile [hunk 3]
```diff
@@ -66,9 +60,8 @@ GOTESTSUM_JSONFILE ?= gotestsum.json
 MMCTL_BUILD_TAGS =
 MMCTL_TESTFLAGS = -timeout 30m -race
 MMCTL_PKG = github.com/mattermost/mattermost/server/v8/cmd/mmctl/commands
-LDFLAGS += -X "$(MMCTL_PKG).gitCommit=$(BUILD_HASH)"
-LDFLAGS += -X "$(MMCTL_PKG).gitTreeState=$(GIT_TREESTATE)"
-LDFLAGS += -X "$(MMCTL_PKG).buildDate=$(BUILD_DATE)"
+MMCTL_BUILD_DATE = $(shell date -u +'%Y-%m-%dT%H:%M:%SZ')
```
Comment (hanzei) [inline server/Makefile:63]: I've added this env variable specifically for mmctl as the server uses a different format that is not standardised.  Maybe it's worth resisting the format used in the server.

**Review by agnivade — APPROVED** [review]
Thanks Ben

**Comment by hanzei** [comment]
@DHaussermann What are the next steps to getting this PR merged?

**Comment by agnivade** [comment]
Ben, I would be fine to merge this if we can add a unit test for the basic case.

**Comment by hanzei** [comment]
@agnivade I did add a very basic unit test in [`5b7afe3` (#23794)](https://github.com/mattermost/mattermost/pull/23794/commits/5b7afe39ff54eb8ddbfa0cb67b4cd51cdcb60e3e). But keep in kind that most the interested data in the version cmd isn't present in unit test as it only exists in binaries as you pointed out earlier.

File: server/cmd/mmctl/commands/version_test.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	s.Require().Len(printer.GetLines(), 1)
+	line := printer.GetLines()[0]
+	s.Require().Contains(line, "mmctl:")
+	s.Require().Contains(line, "Version:")
+	s.Require().Contains(line, "BuiltDate:")
+	s.Require().Contains(line, "CommitDate:")
+	s.Require().Contains(line, "GitTreeState:")
+	s.Require().Contains(line, "GoVersion:")
+	s.Require().Contains(line, "	go1.")
+	s.Require().Contains(line, "Compiler:")
```
Comment (agnivade) [inline server/cmd/mmctl/commands/version_test.go:27]: Compiler should be `gc` always right?

File: server/cmd/mmctl/commands/version_test.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	printer.SetFormat(printer.FormatPlain)
+
+	err := versionCmdF(&cobra.Command{}, []string{})
+	s.Require().NoError(err)
+	s.Require().Len(printer.GetErrorLines(), 0)
+	s.Require().Len(printer.GetLines(), 1)
+	line := printer.GetLines()[0]
+	s.Require().Contains(line, "mmctl:")
+	s.Require().Contains(line, "Version:")
+	s.Require().Contains(line, "BuiltDate:")
```
Comment (agnivade) [inline server/cmd/mmctl/commands/version_test.go:22]: Won't this be the current date always?

**Review by agnivade — COMMENTED** [review]
Some minor comments on hardening a few fields of the tests.

**Comment by hanzei** [comment]
@agnivade The output of `versionCmdF` when running it as part of a test is
```
mmctl:
        	            	Version:	9.1.0
        	            	BuiltDate:	dev
        	            	CommitDate:	dev
        	            	GitCommit:	dev
        	            	GitTreeState:	dev
        	            	GoVersion:	go1.21.0
        	            	Compiler:	
        	            	Platform:	/
```
which is not very helpful for testing. In comparison, ` go run ./cmd/mmctl version` outputs
```
mmctl:
Version:	9.1.0
BuiltDate:	dev
CommitDate:	dev
GitCommit:	dev
GitTreeState:	dev
GoVersion:	go1.21.0
Compiler:	gc
Platform:	amd64/linux
```
and `go install ./cmd/mmctl && mmctl version` provides (almost) the full output:
```
mmctl:
Version:	9.1.0
BuiltDate:	dev
CommitDate:	2023-10-09T11:00:32Z
GitCommit:	a1014f5ec552f26d361f1d602bdf52ff5fa92d99
GitTreeState:	dirty
GoVersion:	go1.21.0
Compiler:	gc
Platform:	amd64/linux
```

Long story short, `debug.ReadBuildInfo()` acts weird in test. It's only content is the go version: `&debug.BuildInfo{GoVersion:"go1.21.0", Path:"", Main:debug.Module{Path:"", Version:"", Sum:"", Replace:(*debug.Module)(nil)}, Deps:[]*debug.Module(nil), Settings:[]debug.BuildSetting(nil)}`

**Comment by agnivade** [comment]
Ok, in that case, I think it'd be good to get QA testing done.

**Comment by DHaussermann** [comment]
@hanzei I gave this a try but don't see it working. 

image

image

I tried this locally where I have Go `1.21.3` The outputs for the command did not show the details I was expecting. I don't see how this would simply not work. I suspect I have misunderstood how I would see this in the output of the shell. 

Can you provide some guidance on how I can see the `debug.BuildInfo`?

**Comment by hanzei** [comment]
@DHaussermann You need also to run the `mmctl` binary that you just built. I've update the QA test steps in the ticket. PTAL.

**Comment by hanzei** [comment]
@DHaussermann The description in the PR is a bit misleading. Running `go install ./cmd/mmctl` and then running the freshly build binary from `$GOBIN/mmctl` should result in the output shown above.

**Comment by DHaussermann** [comment]
Ah thanks @hanzei. Yes I see, it's just running the same binary from a bin folder instead of the working directory.

**Review by DHaussermann — APPROVED** [review]
Tested and passed based on testing above. 

No issues found

LGTM!

**Pull request merged**